### ISSUE TITLE
StringFormat.encodeToString(value: T) rises Error when T is a sealed interface in kotlin 1.6.20-M1 jvm/js environments

### ISSUE THREAD

**Issue by kuzuha** [Kotlin/kotlinx.serialization#1869]
**Describe the bug**
[KClass<T>.serializerOrNull() in kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization/blob/fc9343f06c5184d51df9ad1006d26c60c3230c2a/core/commonMain/src/kotlinx/serialization/Serializers.kt#L177) returns null when T is sealed interface.
SomeSealedInterface::class.serializerOrNull() must be same as SomeSealedInterface.serializer().

This bug makes difference result between inline and non-inline functions.

I suspect [here](https://github.com/Kotlin/kotlinx.serialization/blob/43d5f7841fc744b072a636b712e194081456b5ba/core/jvmMain/src/kotlinx/serialization/internal/Platform.kt#L32) is a problem point.

In this code, check jClass.isInterface before invokeSerializerOnCompanion.
However, after support sealed interface in kotlin 1.6.20, interface could be has serialzer() method in companion object.

So that, simple fix is changing order of those.
But it could make another problem. I don't know details.

**To Reproduce**
```kotlin
import kotlinx.serialization.*
import kotlinx.serialization.json.*

@Serializable
sealed class AA

@Serializable
class BB: AA()

@Serializable
class CC: AA()

@Serializable
sealed interface A

@Serializable
class B : A

@Serializable
class C : A

@OptIn(InternalSerializationApi::class)
fun main(args: Array<String>) {
    A::class.serializerOrNull() // null
    A.serializer() // SealedClassSerializer
    println(A.serializer() == A::class.serializerOrNull()) // false

    AA.serializer() == AA::class.serializerOrNull() // true

    println(Json.encodeToString(A.serializer(), B())) // {"type":"B"}
    println(Json.encodeToString<A>(B())) // kotlinx.serialization.SerializationException: Serializer for class 'A' is not found.
}
```

**Expected behavior**
true
{"type":"B"}
{"type":"B"}

**Environment**
 - Kotlin version: 1.6.20-M1
 - Library version: 1.3.2
 - Kotlin platforms: JVM
 - Gradle version: 7.1
 - IDE version (if bug is related to the IDE) [e.g. IntellijIDEA 2019.1, Android Studio 3.4]
 - Other relevant context [e.g. OS version, JRE version, ... ]

**Comment by kuzuha** [comment]
I confirmed same json.encodeToString<A>(B())) problem occurred in js. Not occured in native.

And then, I confirmed`A.serializer() == A::class.serializerOrNull()` is not guaranteed. 
So, I'll improve title of this issue more reasonable.

**Comment by kuzuha** [comment]
refs #1576

**Comment by kuzuha** [comment]
also refs #1754

**Comment by sandwwraith** [comment]
Closed by #1968
